packages/checkout/widgets-lib/src/widgets/sale/hooks/useCurrency.ts

```diff
@@ -0,0 +1,170 @@
+/* eslint-disable @typescript-eslint/naming-convention */
+import { useState, useEffect } from 'react';
+import { Passport } from '@imtbl/passport';
+
+import { Environment } from '@imtbl/config';
+import { TokenFilterTypes } from '@imtbl/checkout-sdk';
+import { providers } from 'ethers';
+import { PRIMARY_SALES_API_BASE_URL } from '../utils/config';
+
+import {
+  ClientConfig,
+  ClientConfigCurrency,
+  ClientConfigCurrencyConversion,
+  SaleWidgetCurrency,
+  SaleWidgetCurrencyType,
+  SignPaymentTypes,
+} from '../types';
+import { useCheckoutClient } from './useCheckoutClient';
+import { sortCurrencies } from '../utils/sortCurrencies';
+
+type ClientConfigResponse = {
+  contract_id: string;
+  currencies: {
+    base: boolean;
+    decimals: number;
+    erc20_address: string;
+    exchange_id: string;
+    name: string;
+  }[];
+  currency_conversion: {
+    [key: string]: {
+      amount: number;
+      name: string;
+      type: string;
+    };
+  };
+};
+
+const toClientConfig = (response: ClientConfigResponse): ClientConfig => ({
+  contractId: response.contract_id,
+  currencies: response.currencies.map((c) => ({
+    ...c,
+    erc20Address: c.erc20_address,
+    exchangeId: c.exchange_id,
+  })),
+  currencyConversion: Object.entries(response.currency_conversion).reduce(
+    (acc, [key, value]) => {
+      acc[key] = {
+        amount: value.amount,
+        name: value.name,
+        type: value.type as SignPaymentTypes,
+      };
+      return acc;
+    },
+    {} as ClientConfigCurrencyConversion,
+  ),
+});
+
+type UseClientConfigParams = {
+  environment: Environment;
+  environmentId: string;
+  passport: Passport | undefined;
+  defaultCurrency?: string;
+};
+
+export const defaultClientConfig: ClientConfig = {
+  contractId: '',
+  currencies: [],
+  currencyConversion: {},
+};
+
+export const useCurrency = ({
+  environment,
+  environmentId,
+  passport,
+  defaultCurrency,
+}: UseClientConfigParams) => {
+  const [selectedCurrency, setSelectedCurrency] = useState<
+  ClientConfigCurrency | undefined
+  >();
+  const [clientConfig, setClientConfig] = useState<ClientConfig>(defaultClientConfig);
+  const [allCurrencies, setAllCurrencies] = useState<SaleWidgetCurrency[]>([]);
+
+  const checkoutClient = useCheckoutClient({ passport, environment });
```

Comment:
We can optimise things here, the checkout instance was already created when the widget instance is created.
Have a look [here](https://github.com/immutable/ts-immutable-sdk/blob/ffd504b90afb04feeb8b08de60ebde8a4dbcbba6/packages/checkout/widgets-lib/src/widgets/BaseWidgetRoot.ts#L22-L22), notice the `BaseWidgetRoot` class provides the checkout instance. so we can pass it down to SaleWidget and then hold it on the Context object.

Reply:
ah yes! good one!